### ISSUE TITLE
Make `Octokit.ApiVersions` interface merging work with GHES additions in TypeScript

### ISSUE THREAD

**Issue by gr2m** [octokit/octokit-next.js#9]
This is a follow up to https://github.com/octokit/octokit-next.js/pull/7

Here is a video summarizing the problem. The file I'm showing is [`tests/ts/ghes-3.1/test.ts`](https://github.com/octokit/octokit-next.js/blob/9/ghes-types/tests/ts/ghes-3.1/test.ts)

https://user-images.githubusercontent.com/39992/128765836-a5a6e15b-b6ec-4d8b-b502-75f1e55a021d.mp4

I also created a [TypeScript playground](https://www.typescriptlang.org/play?#code/PTAEBUAsFNQYwPYDsAu0AeLQDMECdQVIBLAZ1AAcAbAQwE8BzPBAVyQBNQzRIUULSALhANiRFgCMAdIgC2wBHBQIA1mIVLVYgLRIMKKQCtSwMqRbQTATgCwAKBD2QEGKAAGAN2h5SxZG9AEChQ-JBx8QlcAZQARAGl4ZFIUPBZNAlJIVipOCVgKb1kxNE5lUDzQUmgUJzAy6DEYAgAidgQUOWbAluhUbwo8MmgugAoVaDoEbFBYuKkAQQAFAEkASilQACEaKs5kSNgqrC8fUIAaWoPCOgLyXAI3OCpiXoM8aABHC2SR1YDM7K5WDsYjYbDeV5SWqXKLQWBEWAYCj4EqESxYWTVLJAqgIADu4QIcgoxGeSAYoG8zB8ULs9iRKNASBomNIFBocFgs1AAG97KBKehkXgsMQ+nhsBzYAB5YKhUgAHnAADVvL59vpeuxyONJtNZgsVgA+Xn8gWgE7qpAAfkEEFVp2QAG4zQKaCwiHbkoNyS67ObyjtoABVPBUW2VFJihh+gUAX3sZoZIq44slnNAS2Wpv95raHQQsjtfNzAeaAHEAKLgUDAZrF10BwIqO0SBAIKjQGhIWNNuO9stVmvAfNybTIKh0es5pvm1St9ud7sD839xtr0uU8UDIZ2g1ZgDarXanQAuqAAGQzpsV6u10eF8dISfTvQnFcC2-D179QZVJ8vg2m4BvO5SLl2PaNvGK4bvG9gJnSdjJlgTw7OQswKmaKpqqEgpoBwOoTFMMzxIa2YALygMeBayM0Fy5uAsohEkeFauhpFMfKSoOlaJqUQanFJNxOHIEa9gmiWAqIEg3ppMoeAjEEzEyXajFykkqzXgKRBkFISnyqAlH6UkvYIWaxkqRAgkyX6Zo7HQSBwKA7xfOiIzMB60B2rqxH7isB7YY6SCnppknmu8KAsHgYQ8nGoA7KAizMEUVQKn5ywBTxoSngeKA3NAxEeWgp5GqZ8GJkhQqMvZjk4GwSi4WgPyhWa0nJPAzyvDEJ6FoZTLQASswjGFAqWqEdrUZ09EBu6npUVUcARXRZpxqstm5ki0BKOA+UKjyzYLh2EGgHGRojDQeI0GIHUvKg3U0VILnfCgIxfrWzSrGtSZCltKA7QUe0HWBR3didZ0XVdKGdXdPWyI9nzPa9Q73rDAFTp9vbOAAAigpDaJtSj43g1KgNoVHI8AP54Du-4TlOoBtJYTLtIKZBYPcDOw3Zl3XU8t0oPdcjw65PxvZT25-tAaMfetArOHi+AqOQeIwGERwhOSVyBVa1wFIK1DEHAYiTnabh6IN8QKs0VM08MZ08m4rVJFD-OVhLQx9ebJFxMNjZjcgE025Ly2brNkATQtS3TSdX2IQKBN-bt+2gW2INhKd508y7rxu2g1OS8LiNix9sfxz9227W+3jg1nN05+7VSF25YsPrI0sY99BQVwDyctsDS7p2djYQ7z0MoLnv5DE3osU0HQzt2asdmXY2D1cp+u-f90BKmdHg0FQFiqasdoeAgxCcLF9hAA) with a very much simplified illustration of the problem I'm running into.

The `@octokit-next/*` packages are published from the [`packages/` folders](https://github.com/octokit/octokit-next.js/tree/f376a960a14a12185344446f4cb6d8f89952d0ef/packages). They are written in JavaScript and hand-written `d.ts` files.

When installing `@octokit-next/core`, then the `version` constructor parameter can only be set to `"github.com"` and the only declared REST API endpoint is the root endpoint `GET /`.

--- 

<img width="805" alt="image" src="https://user-images.githubusercontent.com/39992/128758163-0450d1d0-dc7a-4036-a9f9-b89016eb7127.png">

--- 

When installing `@octokit-next/types-rest-api-github.com` and importing it, then the additional endpoints show up

--- 

<img width="824" alt="image" src="https://user-images.githubusercontent.com/39992/128758242-85af2c80-e496-4ea9-81cf-3878ec1c363d.png">

--- 

However, when installing and importing `@octokit-next/types-rest-api-ghes-3.1`, then the `version` constructor option cannot be set to `"ghes-3.1"`, although it is added as key to the `Octokit.ApiVersions` interface:

--- 

<img width="728" alt="image" src="https://user-images.githubusercontent.com/39992/128758408-be712b1e-1074-403f-9d45-0657d8ad7ef3.png">

--- 

https://github.com/octokit/octokit-next.js/blob/f376a960a14a12185344446f4cb6d8f89952d0ef/packages/types-rest-api-ghes-3.1/index.d.ts#L52-L63

--- 

I added tests for both JavaScript (using [`tsd`](https://github.com/SamVerschueren/tsd)) and TypeScript (using `tsc`) examples in [`tests/js`](https://github.com/octokit/octokit-next.js/tree/f376a960a14a12185344446f4cb6d8f89952d0ef/tests/js) and [`tests/ts`](https://github.com/octokit/octokit-next.js/tree/f376a960a14a12185344446f4cb6d8f89952d0ef/tests/ts) respectively. 

~~The JS/DTS tests are passing, but [the TypeScript test for GHES-3.1](https://github.com/octokit/octokit-next.js/blob/f376a960a14a12185344446f4cb6d8f89952d0ef/tests/ts/ghes-3.1/test.ts) is failing. And I'm not sure why that is~~ turns out the the JS/DTS tests used a custom class where I explicitly set the versions. I now added a new [`tests/js/ghes-3.1`](https://github.com/octokit/octokit-next.js/tree/9/ghes-types/tests/js/ghes-3.1) test that fails just like its TypeScript counterpart

**Comment by gr2m** [comment]
I found out that what does work is if I set the `Octokit<TVersion>` type argument explicitly

<img width="791" alt="image" src="https://user-images.githubusercontent.com/39992/128759221-8bddbe9a-1b62-47be-9c2d-a906eb3b73fe.png">

But that should be inferred from the `version` option by default 🤔 

I also realized that the GHES-3.1 tests for JS create a custom class which sets the version type argument explicitly

https://github.com/octokit/octokit-next.js/blob/f376a960a14a12185344446f4cb6d8f89952d0ef/tests/js/ghes-3.1/index.d.ts#L1-L15

While the TS tests does not do that

https://github.com/octokit/octokit-next.js/blob/f376a960a14a12185344446f4cb6d8f89952d0ef/tests/ts/ghes-3.1/test.ts#L1-L18

**Comment by gr2m** [comment]
I think I figured it out. I'll have to remove `version` from `Octokit.Options` and treat it entirely separate. That way it can be derived from the `version` options passed to the constructor.

[playground](https://www.typescriptlang.org/play?#code/PTAEBUAsFNQYwPYDsAu0AeLQDMECdQVIBLAZ1AAcAbAQwE8BzPBAVyQBNQzRIUULSALhANiRFgCMAdIgC2wBHBQIA1mIVLVYgLRIMKKQCtSwMqRbQTATgCwAKBD2QEGKAAGAN2h5SxZG9AEChQ-JBx8QlcAZQARAGl4ZFIUPBZNAlJIVipOCVgKb1kxNE5lUDzQUmgUJzAy6DEYAgAidgQUOWbAluhUbwo8MmgugAoVaDoEbFBYuKkAQQAFAEkASilQACEaKs5kSNgqrC8fUIAaWoPCOgLyXAI3OCpiXoM8aABHC2SR1YDM7K5WDsYjYbDeV5Sez2DAUfBYJA0WSWCg0OCwWagADe9lAoFh8K4fTw2DRsAA8sFQuQcXY8XiaCwiIJKiliEgGABuXH0iQ7aAAVTwVAA-CzkoMOdy6aAAL7QmUEvBYdloElk0BLZbYnl4todBCyFm0+n05oAcQAouBQMBmsbdabAioWRIEAgqNAaEhpU65b6nRbrbb9XJtMgqHR7TqZX7VK73Z7vQHTbKU-7Ha9+oMqizZgsVgBtVrtToAXVAADIY36gzbgKHDeGkJHo3oTum8XXbVm8AMhs3Ww7Y074+VE16fY68WnHbOZfK7IuYeg4cr4LRSORZgAecAANW8vn2+l67HI40m03zWtAAF5QCWDbJmgA+GuJJAStLKPAjIIhEkxqgCcx5ICyB5HqEcpVjM8RSJSgFfqsH54kQZBSAB1L3oEVJJAGy4ylhQHYiBUHIBBh6nPssqwfmiHUtKPI7HQSBwKA7xfJYKAjMwTLQCyl5THBcxaoWkHUUgZYoSa9LvCgLB4GEWK0TsoCLMwRRVDuN5FhJYFloWKA3NAwl8WgZavgR9iEUqWAsWxOBsEo0FoD8Mk8ogX5YE8LyoDEpaGjhegAO4iSMsl4qBoQsk+nQXCOjLMo+VRwPJzQJTOqxMYqq7QEo4AmTuWLOgmHqTnKr4jDQIU0GIG5+SgAXPlInHfDx3Z2qs2U8rC+UoIVBTFaV47ld6lXVbV9W+a8zVyK1nztSMnWNrIg5Rt1AbOAAAigpDaH1SgHXgzAENoj5WvWvb9lU61dG0ligEg7T4ugZBYPcoCrcxU0+c8s2BbIC1cT8nXXTm0B3T1dg8s4IX4Co5AhTAYRHCEHJXPprkma91DEHAYiRiybihSJO7NODQxvhFbieUkf2NZaxI3bAD6k7MEWOtFFGPpTVQZY6SWQLFqXpZlcrQ71eUFUVJVjm6Y1hLKVU1XVDOvEzaos8DS2dc0m1SwUMtDe23gq79DUa8zEM69xy2XSGgNQ9DeKHQNssjQrSZK1VgsWzNqCa9mQy26DDvAHzkMRhtPLQ4R2DOUhuP9YN0B7lVHg0FQFgQasLIeAgxCcCp9hAA)

**Comment by nmkataoka** [comment]
@gr2m Hey I saw your post on the TS discord after you closed it but it looked interesting so I took a crack at it. If you're still looking at this, would something along these lines work?

[playground](https://www.typescriptlang.org/play?#code/PTAEBUAsFNQYwPYDsAu0AeLQDMECdQVIBLAZ1AAcAbAQwE8BzPBAVyQBNQzRIUULSALhANiRFgCMAdIgC2wBHBQIA1mIVLVYgLRIMKKQCtSwMqRbQTATgCwAKBD2QEGKAAGAN2h5SxZG9AEChQ-JBx8QlcAZQARAGl4ZFIUPBZNAlJIVipOCVgKb1kxNE5lUDzQUmgUJzAy6DEYAgAidgQUOWbAluhUbwo8MmgugAoVaDoEbFBYuKkAQQAFAEkASilQACEaKs5kSNgqrC8fUIAaWoPCOgLyXAI3OCpiXoM8aABHC2SR1YDM7K5WDsYjYbDeV5SWqXKLQWBEWAYCj4EqESxYWTVLJAqgIADu4QIcgoxGeSAYoG8zB8ULs9iRKNASBomNIFBocFgs1AAG97KBKehkXgsMQ+nhsBzYAB5YKhUgAHnAADVvL59vpeuxyONJtNZgsVgA+Xn8gWgE7qpAAfkEEFVp2QAG4zQKaCwiHbkoNyS67ObyjtoABVPBUW2VFJihh+gUAX3sZoZIq44slnNAS2Wpv95raHQQsjtfNzAeaAHEAKLgUDAZrF10BwIqO0SBAIKjQGhIWNNuO9stVmvAfNybTIKh0es5pvm1St9ud7sD839xtr0uU8UDIZ2g1ZgDarXanQAuqAAGQzpsV6u10eF8dISfTvQnFcC2-D179QZVJ8vg2m4BvO5SLl2PaNvGK4bvG9gJnSdjJlgTw7OQswKmaKpqqEgpoBwOoTFMMzxIaywXLm4CyiESR4Vq6GkdR8pKg6VpGvYJolgKiBIN6aTKHgIxBDRvHFhaOHIHa2GOmEcaXhATFJKs14CkQZBSMJ8qgAAvIEcpJL2CFmppSRSYpvF+maOx0EgcCgO8XzogqABKrBoHRBGgLqxH7isB7SVap5GiMzAetAdquWFylcea7woCweBhDyck7KAizMEUVQKr5yz+axoSngekVoEFhnwYmSFCoy1m2TgbBKLhaA-NFZo8ck8DPK8MQnoWOlMtABKzCMMUCpaoR2seBayM0FEBu6nqgM0VRwHFM1mnGqyWbmSLQEo4A3NACo8s2C4dhBoBxsFNB4jQYgdS8qDdVNUgOd8KAjF+tbNKsm1JkKu0oPtBRHSdYFnd2F1XTdd1PA9KBPXIL2fG9H1DvePWyABU4-b2zgAAIoKQ2g7UoxN4NSoDaItaPAD+eA7v+E5TqAbSWEy7SCmQWD3CzGNWdDKGdY9GNI45PyfbT25-tAWPfVtArOHi+AqOQeIwGERwhOSVwBY1B2CtQxBwGIk52m4eiDfECrNHTDPDMFPJuK1SSC3DlZS0MfUWyRcTDY2Y2SYttvS2tm7zZAE3Lats0Xb9iECiTgMHSDoFtuDslQ7druvO7aD09LosoxL31xwn-17cnb7eJnMNCygue-kMhfoqjd4jhjss439BQV8Dx2p+BEOXSMjbXVn905x7VTN+LNPB0MndmnHRl2Ng9UiQbANA4d4DBR4NBUBYUmrHaHgIMQnDJfYQA)

**Comment by gr2m** [comment]
Thanks a lot Nolan for the follow up! Setting `options: { version: TVersion } & TOptions` in the constructor is exactly the solution I came up with as well, that way `TVersion` can be derived:

https://github.com/octokit/octokit-next.js/pull/10/files#diff-144904d408ec513806681518c894e534492562fdd352bba77ab8e7643851d978R209

I also removed `version` from `API.Options` to make it explicit everywhere I use it.

The next challenge I see is to make it work when creating custom classes using the static `.withDefaults({ version })` method. Ideally I'd like this to work

```ts
import { Octokit } from "@octokit-next/core"
import "@octokit-next/types-rest-api-ghes-3.1"

const MyOctokit = Octokit.withDefaults({
  version: "ghes-3.1"
})

const octokit = new MyOctokit()
octokit.request("GET /ghes-only")
```

But that's another rabbit hole 🐰🕳️ I created a follow up issue for now #11
